File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.java
// 일관성 없는 전화번호 목록이라도, 일단 입력은 끝까지 받아야한다.
// "1234", "123" 처리만 하고 "123", "1234" 일 때의 처리는 하지 않았다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

class TrieNode {
    Map<Character, TrieNode> childNodes = new HashMap<>();
    boolean isLeaf = false;
}

class Trie {
    TrieNode rootNode;

    public Trie() {
        this.rootNode = new TrieNode();
    }

    void insert(String word) {
        TrieNode node = this.rootNode;

        for (int i = 0; i < word.length(); ++i) {
            node = node.childNodes.computeIfAbsent(word.charAt(i), character -> new TrieNode());
        }

        node.isLeaf = true;
    }

    boolean search(String word) {
        TrieNode node = this.rootNode;

        for (int i = 0; i < word.length(); ++i) {
            node = node.childNodes.getOrDefault(word.charAt(i), null);

            // 중간에 값이 없을 때, 일관성이 있는 전화번호 목록일 때이다.
            // ex. "12340", "12341"
            if (node == null) {
                return true; // 일관성이 있다.
            }

            // 중간 노드가 리프노드로 존재할 시
            // ex. "1234", "123" → '3'일 때 걸림
            if (node.isLeaf) {
                return false; // 일관성이 없다.
            }
        }

        // word의 마지막 노드가 이전에 자식노드가 없을 시, 처음 삽입된 것이므로
        // ex. "123"
        if (node.childNodes.isEmpty()) {
            return true;
        }
        // ex. "123", "1234"
        return false;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine());

        for (int i = 0; i < T; ++i) {
            Trie trie = new Trie();
            boolean consistencyFlag = true;

            int N = Integer.parseInt(br.readLine());
            for (int j = 0; j < N; ++j) {
                String number = br.readLine();

                // 입력값은 일단 계속 받아야 한다.
                if (consistencyFlag == false) {
                    continue;
                }

                // 삽입하기 전에 Trie 자료구조에서 지금 number가 일관성 있는지 체크한다.
                boolean consistency = trie.search(number);

                // 일관성이 있다면 삽입해준다.
                if (consistency) {
                    trie.insert(number);
                }
                // 일관성이 없다면 이후 전화번호는 볼 필요가 없다.
                else {
                    consistencyFlag = false;
                }
            }

            if (consistencyFlag) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
    }
}

File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.py
import sys
input = sys.stdin.readline

class TrieNode:
    def __init__(self, key, isLeaf=None):
        self.key = key
        self.isLeaf = isLeaf
        self.children = {}

class Trie:
    # Trie 헤드 노드 생성
    def __init__(self):
        self.head = TrieNode(None)

    def insert(self, string):
        # 트리 헤드
        curr_node = self.head

        for s in string:
            # 현재 노드에 해당하는 글자의 자식이 없다면 새로운 노드로 추가
            if s not in curr_node.children:
                curr_node.children[s] = TrieNode(s)

            # 다음 노드로 이동
            curr_node = curr_node.children[s]

        # 리프노드에 '전체 문자열' 저장
        curr_node.isLeaf = string


    def search(self, string):
        curr_node = self.head

        for s in string:
            if s in curr_node.children:
                curr_node = curr_node.children[s]
            else:
                return False

        # children이 남아있으면 전화번호가 겹치는 것을 의미한다.
        # 911, 9112 → NO
        if curr_node.children:
            return False

        # 남아있지 않으면 일관성 있는 독립적인 번호이다.
        # 911, 912 → YES
        else:
            return True



T = int(input().strip())
for _ in range(T):
    N = int(input())
    nums = []
    trie = Trie()
    flag = True

    for _ in range(N):
        num = input().strip()
        nums.append(num)
        trie.insert(num)

    # 길이가 짧은 전화번호부터 search를 진행한다.
        # 일관성 체크를 더 빠르게 하기 위함.
    nums.sort()
    for num in nums:
        if not trie.search(num):
            flag = False
            break

    if flag:
        print("YES")
    else:
        print("NO")